tools/clang/lib/AST/HlslTypes.cpp

```diff
@@ -687,6 +687,11 @@ bool GetHLSLSubobjectKind(clang::QualType type,
   hgType = (DXIL::HitGroupType)(-1);
   type = type.getCanonicalType();
   if (const RecordType *RT = type->getAs<RecordType>()) {
+    RecordDecl *RD = RT->getDecl();
+    if (!RD->hasAttr<HLSLSubObjectAttr>()) {
+      return false;
+    }
```

Comment:
Given that you've added the attribute now, it would be pretty easy to avoid the string compares by adding a parameter to it containing an integer representation of the Subobjectkind similar to how HLSLResourceAttr does. The callers of , CreateSubobjectStateObjectConfig and friends that call StartSubObjectDecl has the AR, which indicates which DXIL:SubobjectKind they need. It would be pretty easy to plumb it down and apply it to the attr.